### DISCUSSION TITLE
Polished WInUI3 application cannot be started.

### DISCUSSION THREAD

**Discussion by JozefGula** [microsoft/WindowsAppSDK#3886] [Q&A] [answered]
Hello, I created a WinUI3 application. The publishing of the application is successful, but it is not possible to start this application. Event Viewer shows:

Faulting application name: MyApp.exe, version: 1.0.0.0, time stamp: 0x64e80000
Faulting module name: KERNELBASE.dll, version: 10.0.19041.3393, time stamp: 0x6b4de7c9
Exception code: 0xc0000602
Fault offset: 0x000000000012d9b2
Faulting process id: 0x3e98
Faulting application start time: 0x01d9f209436c3f7e
Faulting application path: E:\Software\MyApp\MyApp\bin\Release\Publish\MyApp.exe
Faulting module path: C:\Windows\System32\KERNELBASE.dll
Report Id: 51fef607-3cde-456e-90d8-b0de1c9b9aa2
Faulting package full name: 
Faulting package-relative application ID:

project:

```
<PropertyGroup>
	<OutputType>WinExe</OutputType>
	<TargetFramework>net7.0-windows10.0.19041.0</TargetFramework>
	<TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
	<RootNamespace>MyApp</RootNamespace>
	<ApplicationManifest>app.manifest</ApplicationManifest>
	<Platforms>x86;x64;</Platforms>
	<RuntimeIdentifiers>win10-x86;win10-x64;</RuntimeIdentifiers>
	<PublishProfile>win10-$(Platform).pubxml</PublishProfile>
	<UseWinUI>true</UseWinUI>
	<EnableMsixTooling>true</EnableMsixTooling>
</PropertyGroup>
```

Publish profile:

```
<PropertyGroup>
	<Configuration>Release</Configuration>
	<Platform>Any CPU</Platform>
	<PublishDir>bin\Release\Publish</PublishDir>
	<PublishProtocol>FileSystem</PublishProtocol>
	<_TargetId>Folder</_TargetId>
	<TargetFramework>net7.0-windows10.0.19041.0</TargetFramework>
	<RuntimeIdentifier>win10-x64</RuntimeIdentifier>
	<SelfContained>false</SelfContained>
	<PublishSingleFile>true</PublishSingleFile>
	<PublishReadyToRun>true</PublishReadyToRun>
</PropertyGroup>
```

==================
Visual Studio Version 17.8.0 Preview 2.0 .... or ..... Version 17.7.4
Os: Windows 10 build 19045.3448

==================

The only thing that doesn't suit me is:
Faulting module path: C:\Windows\**System32**\KERNELBASE.dll. Could it be a problem that it refers to System32? If so, why is it happening? 

Can someone please advise me where and what the problem could be? Of course, the application in debug mode works without problems and errors...

Thank you very much

**Comment by ghost1372** [comment, marked as answer]
turn off PublishSingleFile
this is not supported yet

**Reply by JozefGula** [reply]
I turned off PublishSingleFile however the result is still the same:

Application: MyApp.exe
CoreCLR Version: 7.0.1123.42427
.NET Version: 7.0.11
Description: The process was terminated due to an unhandled exception.
Exception Info: System.DllNotFoundException: Unable to load DLL '**Microsoft.ui.xaml.dll**' or one of its dependencies: The specified module could not be found. (0x8007007E)
   at MyApp.Program.Main(String[] args) in E:\\Software\MyApp\MyApp\obj\x64\Release\net7.0-windows10.0.19041.0\win10-x64\App.g.i.cs:line 28

Faulting application name: MyApp.exe, version: 1.0.0.0, time stamp: 0x64e80000
Faulting module name: KERNELBASE.dll, version: 10.0.19041.3393, time stamp: 0x6b4de7c9
Exception code: 0xe0434352
Fault offset: 0x000000000002cf19
Faulting process id: 0xa20
Faulting application start time: 0x01d9f21419845f2d
Faulting application path: E:\softvare_developer\csharp_solution\Software\MyApp\MyApp\bin\Release\Publish\MyApp.exe
Faulting module path: C:\Windows\System32\KERNELBASE.dll
Report Id: 35f003ec-c9a9-4875-a7c3-d5b9579d1583
Faulting package full name: 
Faulting package-relative application ID: 

Publish profile:

```
  <PropertyGroup>
    <Configuration>Release</Configuration>
    <Platform>Any CPU</Platform>
    <PublishDir>bin\Release\Publish</PublishDir>
    <PublishProtocol>FileSystem</PublishProtocol>
    <_TargetId>Folder</_TargetId>
    <TargetFramework>net7.0-windows10.0.19041.0</TargetFramework>
    <RuntimeIdentifier>win10-x64</RuntimeIdentifier>
    <SelfContained>false</SelfContained>
    <PublishSingleFile>false</PublishSingleFile>
    <PublishReadyToRun>true</PublishReadyToRun>
  </PropertyGroup>
```

Thanks a lot

**Reply by ghost1372** [reply]
add this line to your csproj and propertygroup
`<WindowsAppSDKSelfContained>true</WindowsAppSDKSelfContained>`

**Reply by JozefGula** [reply]
Hello, I edited the project. Event Viewer no longer shows the problem with "Microsoft.ui.xaml.dll", but there is still a problem with KERNELBASE.dll.

Event Viewer:

Faulting application name: MyApp.exe, version: 1.0.0.0, time stamp: 0x64e80000
Faulting module name: KERNELBASE.dll, version: 10.0.19041.3393, time stamp: 0x6b4de7c9
Exception code: 0xc000027b
Fault offset: 0x000000000012d9b2
Faulting process id: 0x41b8
Faulting application start time: 0x01d9f2177afab7a3
Faulting application path: E:\Software\MyApp\Pedos\bin\Release\Publish\MyApp.exe
Faulting module path: C:\Windows\System32\KERNELBASE.dll
Report Id: 2342ac01-c3aa-4133-9324-c9a36452150c
Faulting package full name: 
Faulting package-relative application ID: 

Project:

```
<PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net7.0-windows10.0.19041.0</TargetFramework>
    <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
    <RootNamespace>Pedos</RootNamespace>
    <ApplicationManifest>app.manifest</ApplicationManifest>
    <Platforms>x86;x64;</Platforms>
    <RuntimeIdentifiers>win10-x86;win10-x64;</RuntimeIdentifiers>
    <PublishProfile>win10-$(Platform).pubxml</PublishProfile>
    <WindowsAppSDKSelfContained>true</WindowsAppSDKSelfContained>
    <UseWinUI>true</UseWinUI>
    <EnableMsixTooling>true</EnableMsixTooling>
</PropertyGroup>
```

Publish profile:

```
<PropertyGroup>
	<Configuration>Release</Configuration>
	<Platform>Any CPU</Platform>
	<PublishDir>bin\Release\Publish</PublishDir>
	<PublishProtocol>FileSystem</PublishProtocol>
	<_TargetId>Folder</_TargetId>
	<TargetFramework>net7.0-windows10.0.19041.0</TargetFramework>
	<RuntimeIdentifier>win10-x64</RuntimeIdentifier>
	<SelfContained>false</SelfContained>
	<PublishSingleFile>false</PublishSingleFile>
	<PublishReadyToRun>true</PublishReadyToRun>
</PropertyGroup>
```

Is PublishReadyToRun supported and I can set it to true?

Thanks a lot

**Reply by ghost1372** [reply]
yes publishreadytorun is supported i use it in my  unpackaged app
https://github.com/WinUICommunity/TvTime/blob/main/dev/TvTime.csproj

however i dont know does this supported in packaged mode or not?

**Reply by JozefGula** [reply]
If I may ask, does Publish work in your project? Can you run this application? I do not see the profile anywhere. It is possible that I am wrong...

* I can not start your project because I can not find "CommunityToolkit.Labs.WinUI.TokenView" anywhere.

**Reply by ghost1372** [reply]
you should add some nuget feed first:
`https://pkgs.dev.azure.com/dotnet/CommunityToolkit/_packaging/CommunityToolkit-MainLatest/nuget/v3/index.json
`
and

`https://pkgs.dev.azure.com/dotnet/CommunityToolkit/_packaging/CommunityToolkit-Labs/nuget/v3/index.json`

oh i see that i forget to push publish profile, however if you see my action in github here
https://github.com/WinUICommunity/TvTime/blob/6de7b4942651bc493e74a92b674f437e2628c27a/.github/workflows/publish.yml#L30
github can build and publish my app into release section.

**Reply by JozefGula** [reply]
Your application successfully fights with me :)

![image](https://github.com/microsoft/WindowsAppSDK/assets/8478191/13962462-7494-4e79-af9f-884db12a1296)

I tried to use your command for my application, but I still can not start the application and it ends up with an error in KERNELBASE.dll

dotnet publish MyApp.csproj -c Release -r win10-x64 -p:Platform=x64 --self-contained true -p:PublishSingleFile=false -p:PublishReadyToRun=true -p:PublishTrimmed=true -p:TrimMode=partial --output publish

**Reply by ghost1372** [reply]
this is community package bugs
however i dont have any issues. i and github actions can compile codes successfully
give me your project, this is something with your visual studio and machine, try to install latest wasdk runtimes
https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/downloads

**Reply by JozefGula** [reply]
Hi, I tried to create a new empty "Blank App" project and tried to publish it with your command. This application is working after publishing. 

Hmm, I will try to compare the projects and all their parts to see if I can find any difference.
If I can't find anything, I will be happy to show you my whole project. Please excuse me, but it is already evening here and I will continue tomorrow because of the children. I will write to you within 24 hours. Thank you very much for your willingness and patience :)

**Reply by ghost1372** [reply]
You can also see my other project, I executed the publish action for it just a few moments ago.
And you can see that it has been successfully executed
https://github.com/ghost1372/ArtWork/actions/workflows/Publish.yml

https://github.com/ghost1372/ArtWork/actions/runs/6341583790/job/17225545014

**Reply by ghost1372** [reply]
Feel free I will help as much as I can

**Reply by JozefGula** [reply]
Hello @ghost1372, I finally managed to find the cause of my problem. I found that your application "ArtWork" works even after publishing. So it seemed that the problem was not Visual Studio, or my environment, but rather my project/application.

I finally found that my application failed on this line of code:

`configHost.SetBasePath(Package.Current.InstalledLocation.Path);`

**System.InvalidOperationException: No such interface supported**

Since the error occurred during the creation of IHost, nothing caught it, but as usual the error was not in WinUI3 but between the keyboard and the screen :(.

Amazingly, the problem is solved, mainly because of you, because you guided me in the right direction. I don't know how to thank you because I spent 3 days on this and without you and your "ArtWork" I would really think that the problem is in WinUI3.

======================

However, allow me one more question. My application is working properly, however, all assets in the published application are missing. All my assets have "Build Action" set to "Content" and are displayed in debug or release mode from Visual Studio. However, they are not displayed only in the published application. Can not you think of a reason?

Thank you very much

**Reply by ghost1372** [reply]
I am glad that your issue is fixed (Package.Current is not supported in UnPackaged app)
your solution is in artwork project again!

just copy and paste this code into csproj:

```
	<ItemGroup>
		<None Include="Assets\**\*">
			<CopyToOutputDirectory>Always</CopyToOutputDirectory>
		</None>
	</ItemGroup>
```

This code copies all the content of the Assets to the output location of the app.

you dont need to set BuildAction anymore.

also pay attention to my csproj i removed all elements like:
<Remove ..../>
<Include..../>

https://github.com/ghost1372/ArtWork/blob/main/dev/ArtWork.csproj

**Reply by JozefGula** [reply]
@ghost1372 thank you very much for your time and willingness. I also read your "ArtWork" code and there is a lot of inspiration on how I can improve my code and its structure. 

**I bow before you**. Thank you very much again.

**Reply by ghost1372** [reply]
your welcome😅
If you are interested in WinUI like me, you can also look at these projects, there are many interesting things
https://github.com/WinUICommunity/WinUICommunity
https://github.com/WinUICommunity/TvTime
https://ghost1372.github.io/

**Reply by JozefGula** [reply]
Thanks, I will definitely look into it.
